src/internal/operators/throwIfEmpty.ts

```diff
@@ -16,32 +18,57 @@ import { MonoTypeOperatorFunction } from '../types';
  *
  * const click$ = fromEvent(button, 'click');
  *
- * clicks$.pipe(
+ * click$.pipe(
  *   takeUntil(timer(1000)),
  *   throwIfEmpty(
  *     () => new Error('the button was not clicked within 1 second')
  *   ),
  * )
  * .subscribe({
  *   next() { console.log('The button was clicked'); },
- *   error(err) { console.error(err); },
+ *   error(err) { console.error(err); }
  * });
  * ```
  *
  * @param {Function} [errorFactory] A factory function called to produce the
  * error to be thrown when the source observable completes without emitting a
  * value.
  */
-export const throwIfEmpty =
-  <T>(errorFactory: (() => any) = defaultErrorFactory) => tap<T>({
-    hasValue: false,
-    next() { this.hasValue = true; },
-    complete() {
-      if (!this.hasValue) {
-        throw errorFactory();
-      }
+export function throwIfEmpty <T>(errorFactory: (() => any) = defaultErrorFactory): MonoTypeOperatorFunction<T> {
+  return (source: Observable<T>) => {
+    return source.lift(new ThrowIfEmptyOperator(errorFactory));
+  };
+}
+
+class ThrowIfEmptyOperator<T> implements Operator<T, T> {
+  constructor(private errorFactory: () => any) {
+  }
+
+  call(subscriber: Subscriber<T>, source: any): TeardownLogic {
+    return source.subscribe(new ThrowIfEmptySubscriber(subscriber, this.errorFactory));
+  }
+}
+
+class ThrowIfEmptySubscriber<T> extends Subscriber<T> {
+  private hasValue: boolean = false;
+
+  constructor(destination: Subscriber<T>, private errorFactory: () => any) {
+    super(destination);
+  }
+
+  protected _next(value: T): void {
+    this.hasValue = true;
+    this.destination.next(value);
+  }
+
+  protected _complete() {
+    if (!this.hasValue) {
+        this.destination.error(this.errorFactory());
```

Comment:
this is going to sound funny, but since `errorFactory` is a _user provided function_, we need to wrap it in a `try/catch`, and send the error along to `destination.error` if there is one. And you can't do it along with the other `destination.error` call, or any error that happens in there will unwind back to your try catch... so it has to be like:

```ts
if (!this.hasValue) {
  let err: any;
  try {
    err = this.errorFactory();
  } catch (e) {
    err = e;
  }
  this.destination.error(err);
} else {
```

Reply:
Makes sense... I will fix it. Thanks for the review!